Update CFileHelper::removeDirectory()

WadeShuler opened the pull request:
Swapped function with a more efficient and reliable way to delete an entire folder and it's contents recursively.

samdark commented:
Is it OK with symlinks?

samdark commented:
btw., it's not compabile with previous method because of signuate change.

WadeShuler commented:
I have replaced it in my framework and it works great. I haven't tested symlinks, but you delete them the same way you do files, using unlink. The function checks if the target is a dir, else it uses unlink. So it should work fine.

I don't know what your talking about with "method signature change"? Are you referring to the params used for calling the function? That could easily be fixed in 2 seconds.....

samdark commented:
Yes, I'm referring to params.

WadeShuler commented:
That is a simple fix, and I honestly didn't think about it breaking people's apps when rolled out lol. I could easily (or anyone could) fix the params.

WadeShuler commented:
Added the $options param (set to null by default so it isn't required).

I edited the file on my branch for the pull request. Do I need to do anything else? I am still new to public coding, and Git. I use Git locally myself, just new to the whole sharing and public aspect to it. Just let me know :)

samdark commented:
Well, not only params. It was possible to disable deleting stuff from symlinked dirs. Now it's not.

WadeShuler commented:
Gotcha. Maybe the original function should be modified. Replacing the
actual deletion process with mine, and options handling from the original.
5 mins and see if this will solve it.

On Sat, Jan 24, 2015 at 3:39 PM, Alexander Makarov <notifications@github.com
> wrote:

> Well, not only params. It was possible to disable deleting stuff from
> symlinked dirs. Now it's not.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/yiisoft/yii/pull/3690#issuecomment-71336189>.
>

WadeShuler commented:
Updated and tested it a bit. Seems to produce the same results as the original. Using scandir is a better way than using glob.

NOTE: Using the original Yii removeDirectory function, it worked JUST fine for me on Windows (7 and 8). I ran into problems with the function when I moved the project onto a Linux server for further testing. So the issue I had was on Linux, and not on Windows. Linux was CentOS 6.6 64b btw.

I already replaced the function in the framework on the Linux server, and in my project. So for me, it's been running live for a while. Seemed to solve the problem for me :)

samdark commented on framework/utils/CFileHelper.php:
```diff
@@ -75,32 +75,25 @@ public static function copyDirectory($src,$dst,$options=array())
-			else
-				rmdir($directory);
 		}
+		return rmdir($directory);
```
Why there's now no `unlink` for directory? It was introduced since in Windows environment symlinks are a bit different than on Windows.

WadeShuler commented:
C:\wamp\www\TEST\deleteme
 + testdir1
   - symdeleteme2
   + directory
      - file.zip
   - file.txt
 + testdir2
   + directory
      - file.doc
   - file.txt
 - file.txt
 - file.csv
 - file.zip

C:\Users\Name\Desktop\deleteme2
  - testdir1
    - file.txt
  - testdir2
    - file.txt


I had a folder 'deleteme' with a few file types, and a few dirs. Inside each dir, had a few more files. In one of them, I had created a symlink to a folder located on my desktop, 'deleteme2'. That folder on my desktop, had files and folders in it too.

WIth traversesymlinks off or omitted, it deleted the 'deleteme' folder. So it deleted it's contents, the symlink, everything. The deleteme2 folder on the Desktop was untouched.

With traverse links set to true, it also deleted the contents of the symlink directory ('deleteme2' on my Desktop, with files and folders in it). The 'deleteme2' directory itself was still on the Desktop, but it was empty.

The only thing I noticed, and was unsure of, was 'deleteme2' (the folder on my desktop) was empty but not deleted itself. So I used the original yii function, and it produced the same results. Everything everywhere was deleted, but the folder of the symlink (the folder on my Desktop) was still there, but it was emptied.

Is the directory the symlink points to supposed to be deleted as well? If so, the original Yii function doesn't do it either.

This was all on Windows. I can further test on Linux in a bit.

samdark commented:
OK. Will check it one more time but now it looks much better.

ua2004 commented:
I tried the original removeDirectory method and the one suggested by @WadeShuler but they both throw a warning: *"rmdir(...): Permission denied"*. All the files inside are removed though. Except the folder itself.

samdark commented:
Won't change current implementation.
